crates/bevy_ecs/src/observer/mod.rs

```diff
@@ -0,0 +1,599 @@
+    }
+
+    #[test]
+    fn observer_order_insert_remove() {
+        let mut world = World::new();
+        world.init_resource::<R>();
+
+        world.observer(|_: Observer<OnAdd, A>, mut res: ResMut<R>| res.assert_order(0));
+        world.observer(|_: Observer<OnInsert, A>, mut res: ResMut<R>| res.assert_order(1));
+        world.observer(|_: Observer<OnRemove, A>, mut res: ResMut<R>| res.assert_order(2));
+
+        let mut entity = world.spawn_empty();
+        entity.insert(A);
+        entity.remove::<A>();
+        entity.flush();
+        assert_eq!(3, world.resource::<R>().0);
+    }
+
+    #[test]
+    fn observer_order_recursive() {
+        let mut world = World::new();
+        world.init_resource::<R>();
+        world.observer(
+            |obs: Observer<OnAdd, A>, mut res: ResMut<R>, mut commands: Commands| {
+                res.assert_order(0);
+                commands.entity(obs.source()).insert(B);
+            },
+        );
+        world.observer(
+            |obs: Observer<OnRemove, A>, mut res: ResMut<R>, mut commands: Commands| {
+                res.assert_order(2);
+                commands.entity(obs.source()).remove::<B>();
+            },
+        );
+
+        world.observer(
+            |obs: Observer<OnAdd, B>, mut res: ResMut<R>, mut commands: Commands| {
+                res.assert_order(1);
+                commands.entity(obs.source()).remove::<A>();
+            },
+        );
+        world.observer(|_: Observer<OnRemove, B>, mut res: ResMut<R>| {
+            res.assert_order(3);
+        });
+
+        let entity = world.spawn(A).flush();
+        let entity = world.get_entity(entity).unwrap();
+        assert!(!entity.contains::<A>());
+        assert!(!entity.contains::<B>());
+        assert_eq!(4, world.resource::<R>().0);
+    }
+
+    #[test]
+    fn observer_multiple_listeners() {
+        let mut world = World::new();
+        world.init_resource::<R>();
+
+        world.observer(|_: Observer<OnAdd, A>, mut res: ResMut<R>| res.0 += 1);
+        world.observer(|_: Observer<OnAdd, A>, mut res: ResMut<R>| res.0 += 1);
+
+        world.spawn(A).flush();
+        assert_eq!(2, world.resource::<R>().0);
+    }
+
+    #[test]
+    fn observer_multiple_events() {
+        let mut world = World::new();
+        world.init_resource::<R>();
+        world.init_component::<A>();
+
+        world
+            .observer_builder::<OnAdd>()
+            .on_event::<OnRemove>()
+            .run(|_: Observer<_, A>, mut res: ResMut<R>| res.0 += 1);
+
+        let entity = world.spawn(A).id();
+        world.despawn(entity);
+        assert_eq!(2, world.resource::<R>().0);
+    }
+
+    #[test]
+    fn observer_multiple_components() {
+        let mut world = World::new();
+        world.init_resource::<R>();
+        world.init_component::<A>();
+        world.init_component::<B>();
+
+        world.observer(|_: Observer<OnAdd, (A, B)>, mut res: ResMut<R>| res.0 += 1);
```

Comment:
No there's no current way to listen to only events where they are both added, if you wanted that behavior you can just listen to one and check it against a query.

Reply:
You mean listen to `Observer<OnAdd, A>` and then in the observer system run a query such as `Query<Entity, Added<B>>`?

Reply:
I'm considering "merging" queries and observers again like I had in the initial implementation to make expressing this easier.

Reply:
How come I wouldn't be able to ensure that they were both just added? The `Added` query doesn't work in side the observer system?

Reply:
The problem is the of existence of user implementations of WorldQuery make it really unclear which components you are listening to based on a `QueryData` function signature. I feel like I'll have to leave that until after a hypothetical query v2.

Reply:
Sorry I keep replying before your message shows up for me which strangely interleaves the conversation and looks like I'm trying to respond to your questions. `Added` theoretically could work in an observer system but that would only signify that the component was added since the last time the observer was run, not that the component was added as part of this event.

Reply:
I see; thanks.
It doesn't have to be part of the v1, but reacting on the Insertion of multiple components together is definitely something that could be useful.
I was actually thinking of switching to it in lightyear; instead of creating an event `ReceivedComponent<C>` for each component that got replicated over the network, I could let users directly add observers `Observers<OnAdd, (Replicate, C)>` or even `Observers<OnAdd, (Replicate, C1, C2, ...)>` to be sure to target the correct entity